src/types/mod.rs

```diff
@@ -217,6 +241,29 @@ impl Type {
         }
     }
 
+    pub fn priority(&self, cache: &ModuleCache<'_>) -> TypePriority {
+        use Type::*;
+        match self {
+            Primitive(_) | UserDefined(_) | Struct(_, _) | Tag(_) => TypePriority::MAX,
+            TypeVariable(id) => match &cache.type_bindings[id.0] {
+                TypeBinding::Bound(typ) => typ.priority(cache),
+                TypeBinding::Unbound(..) => TypePriority::MAX,
+            },
+            Function(_) => TypePriority::FUN,
+            TypeApplication(ctor, _) if ctor.is_polymorphic_int_type() => TypePriority::MAX,
+            TypeApplication(ctor, _) if ctor.is_polymorphic_float_type() => TypePriority::MAX,
```

Comment:
```suggestion
```
These cases should be removed I think. I haven't looked at it in a bit but I'm assuming these must have been included because of an existing parsing issue or similar but the intention was for `Int a` to parse like any other type application.

Reply:
Yes.
These branches give special treatment to polymorphic numeric types. If these branches were simply removed, the representation of concrete types such as `I32` would always be accompanied by parentheses, as in `(I32)`.

```
// removed `TypeApplication(ctor, _) if ctor.is_polymorphic_int_type() => TypePriority::MAX,`
examples/typechecking/bind.an: Actual stdout differs from expected stdout:
  1| add_one : forall a. (Maybe I32 -> Maybe I32 can a)
   | add_one : forall a. (Maybe (I32) -> Maybe (I32) can a)
..
  4| x : Maybe I32
   | x : Maybe (I32)
```

Is there any way to properly distinguish between these?

Reply:
Oh, good question! Yes, you can distinguish between them. So when you have `Int I32` we try to display it as just `I32` but internally it will still be `TypeApplication(Type::Primitive(PrimitiveType::IntegerType), vec![i32 type])`. So you can check for that case specifically by checking what the type variable `a` is bound to in `Int a` (if it is bound).

You can find code that checks for this case in `fmt_type_application` - the `is_polymorphic_int_type` check on the constructor. Then you can see how `fmt_polymorphic_numeral` afterward checks the type bindings on the first argument to see whether it is a type variable (unbound) or already bound to a type.

Reply:
The point raised have been corrected 💪